### PR TITLE
lib: Provide library to support sql data retention sites (closes #438)

### LINKED ISSUE #438: Lib - Postgresql - Provide library to support postgresql as data retention site
Postgresql is one of the most widely used relational database management systems, so it's a natural need to provide support using it within the EDC.

<sub> Denis Neuling <denis.neuling@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)

### PR DESCRIPTION
**Pull request by denisneuling**
This PR is an alternative implementation to [PR 439](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/439) to support SQL databases as data retention site.

According to [Pauls comment](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/439#pullrequestreview-836662004) following changes have been introduced:
* Implementation has been generalized to be applicable to various sql backends (e.g. derby, h2 etc) 
* Usage of [Testcontainers](https://www.testcontainers.org/) has been replaced by an explicit step in the gh actions workflow file
* Transaction management has been dropped for complexitiy reasons
* Connection pooling using [commons-pool2](https://commons.apache.org/proper/commons-pool/) has been decoupled from the sql-lib thus made an optional feature
* postgres specific code has been isolated into its own gradle module

@ndr-brt, @paullatzelsperger would you like to give this another review shot?

<sub> Denis Neuling <denis.neuling@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)

### CODE REVIEW

File: .github/workflows/integrationtests.yaml [hunk 1]
```diff
@@ -107,4 +107,31 @@ jobs:
...
-        run: ./gradlew clean extensions:iam:daps:check
\ No newline at end of file
+        run: ./gradlew clean extensions:iam:daps:check
+
+  Lib-Common-SQL-Postgresql-Integration-Test:
+    runs-on: ubuntu-latest
+
+    services:
+      postgres:
+        image: postgres:9.6.12
```
[the remark this answers is not in the document]
Comment (denisneuling) [reply .github/workflows/integrationtests.yaml]: makes absolutely sense. using postgres:14.1 now.

File: common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/SqlQueryExecutor.java [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+    /**
+     * Intended for mutating queries.
+     *
+     * @param sql       the parametrized sql query
+     * @param arguments the parameteres to interpolate with the parametrized sql query
+     * @return rowsChanged
+     * @throws SQLException if execution of the query was failing
+     */
+    public static int execute(Connection connection, String sql, Object... arguments) throws SQLException {
```
[the remark this answers is not in the document]
Comment (denisneuling) [reply common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/SqlQueryExecutor.java]: The `SqlQueryExecutor` shall neither open nor close any connections, that's rather a task of e.g. a connection pool resp. obligation of the caller. If we would pass the connection via the constructor one would have to instantiate a new `SqlQueryExecutor` instance whenever a query is issued; afterwards the instance is right away going to be garbage collected: Seems to me somewhat superfluous. Could also remove that class if those utility methods aren't needed?

File: common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/SqlQueryExecutor.java [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        if (resultSet != null) {
+            while (resultSet.next()) {
+                results.add(resultSetMapper.mapResultSet(resultSet));
+            }
+        }
+
+        return results;
+    }
+
+    private interface ArgumentHandler {
```
[the remark this answers is not in the document]
Comment (denisneuling) [reply common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/SqlQueryExecutor.java]: Makes sense. `ArgumentHandler` and `ArgumentHandlers` have been extracted to package private classes

File: common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/SqlQueryExecutor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ndr-brt) [reply common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/SqlQueryExecutor.java]: Yes, I was thinking about more on having a `SQLClient` object that could take care of executing queries and statement hiding details about connection retrieval and stuff, but it can be build eventually afterwards on top of the executor. 
So this is ok for me :+1:

**Review by ndr-brt — APPROVED** [review]

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]
First of all this looks A LOT cleaner than before, so great job! 

A few comments here :
- In order to conform to the current extension scheme the code located in `common/libraries` should go into `extensions/sql/[common|pool|postgres]`. I know that there is an ongoing discussion (Issue #92) but until this has been finally decided, we should stick to what we have, so as not to confuse readers.

Regarding the TransactionManagement: 
The way it is now we're operating on 'raw' connections, leveraging their `commit()` and `rollback()` methods. This is fine for local transactions, i.e. transactions against a single database that can be managed directly by client code.

However, there will be use cases where we will require managed transaction (e.g. embedded in a Spring app, or when using XA). Those are transactions, that involve several data sources and are managed by a transaction manager.

Therefore we need some sort of abstraction, which we called the `TransactionContext` in the previous PR, for which there would then be several implementations: 

1) A `ResourceLocalTransactionContext` which simply controls transactions using `Connection.commit()` and `Connection.rollback()` (or even using the auto-commit feature).

2) A `DistributedTransactionContext` which does **NOT** use `commit()` and `rollback()`, but instead use e.g. `javax.transaction.UserTransaction` to demarcate the transaction, depending on what TA is actually used.

We may not need a `DistributedTransactionContext` now, and to complete this PR, but we should implement the infrastructure for it and the `ResourceLocalTransactionContext`, so as not to break client code in the future. I'm happy to help.

File: .github/workflows/integrationtests.yaml [hunk 4]
```diff
@@ -107,4 +107,31 @@ jobs:
...
+        uses: actions/setup-java@v1
+        with:
+          java-version: '11'
+      - name: AWS S3 StatusChecker Test
+        env:
+          RUN_INTEGRATION_TEST: true
+          POSTGRES_PASSWORD: pgpass
+          POSTGRES_USER: pguser
+          POSTGRES_DB: pgdb
+        run: ./gradlew :common:libraries:sql-postgresql-lib:check
```
[the remark this answers is not in the document]
Comment (denisneuling) [reply .github/workflows/integrationtests.yaml:139]: Thx for hinting me towards this. Has been fixed and amended.

File: common/libraries/sql-postgresql-lib/src/test/java/org/eclipse/dataspaceconnector/sql/postgresql/PostgresqlConnectionFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+import java.util.UUID;
+
+class PostgresqlConnectionFactoryTest {
+    private static final String JDBC_URI_STRING = "jdbc://uri-is-mandatory";
+    private static final URI JDBC_URI = URI.create(JDBC_URI_STRING);
+    private static final String USERNAME = "username is mandatory";
+
+    @BeforeAll
+    static void setUp() {
+        Mockito.mockStatic(DriverManager.class);
```
[the remark this answers is not in the document]
Comment (denisneuling) [reply common/libraries/sql-postgresql-lib/src/test/java/org/eclipse/dataspaceconnector/sql/postgresql/PostgresqlConnectionFactoryTest.java]: Added some invocation verification

**Comment by denisneuling** [comment]
@paullatzelsperger As agreed offline `java.sql.DataSource` implementations have been added. 
Not sure whether to move these gradle sql-* modules to the extensions module, since the modules themselves provide rather utility functionality. Do you still insist on moving them? 
Also, would appreciate, if you'd have another look onto the added classes.

**Comment by paullatzelsperger** [comment]
sorry for having been slow on this, will re-review and answer until today EOB.

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]
In order to remain consistent with how extensions are structured now, I would suggest the following change to the structure:
- `common:libraries:sql-lib` --> `extensions:sql:sql-common`: contains "library code"
- `common:libraries:sql-pool...` --> `extensions:sql:pool:apache-commons-pool`: contains a pool impl. (most Tx managers bring their own)
- `common:libraries:sql-postgres...` --> `extensions:sql:postgres`: contains postgres specific configuration. _We will be able to remove this altogether, once PR #561 is merged._

And in the future:
- `extensions:sql:assetindex-sql`
- `extensions:sql:contract-negotiation-store-sql` and so forth. 

Assuming that they are independent of the actual SQL dialect (postgres, mariadb, etc.).

Whether or not that is the perfect structure is a matter of discussion, but it should at least be consistent with the rest.

Lastly, I think that the Apache Commons Pool extension should register the datasource that it contributes in the `DataSourceRegistry`. That way adding the dependency to a build file automatically makes the pooled data source available.

File: common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/ResultSetMapper.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+public interface ResultSetMapper<T> {
+
+    /**
+     * Maps a sql result set into a object.
+     *
+     * @param resultSet containing columns of a row
+     * @return result
+     * @throws SQLException if something went wrong
+     */
+    T mapResultSet(ResultSet resultSet) throws SQLException;
```
[the remark this answers is not in the document]
Comment (denisneuling) [reply common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/ResultSetMapper.java:35]: Must disagree, this is utility functionality and not part of any service extension. The integrating service extension has to handle those exception.

[the remark this answers is not in the document]
Comment (paullatzelsperger) [reply common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/ResultSetMapper.java]: The `ResultSetMapper` is only used in the `SqlQueryExecutor`, which will in turn likely be used in `*store` implementations, where try-catching the same SQL exception is nothing but a nuisance, because they will be rethrown as `EdcExceptions` anyway, so we'd only introduce multiple identical try-catch blocks.

On a more general note: checked exceptions should be wrapped in an unchecked exception as soon as possible. I know checked exceptions are an ambivalent topic, but a design decision has been made to not use them.

File: extensions/transaction/transaction-local/src/main/java/org/eclipse/dataspaceconnector/transaction/local/LocalDataSourceRegistry.java [hunk 7]
```diff
@@ -14,36 +14,41 @@
...
 import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiConsumer;
 import javax.sql.DataSource;
 
 /**
  * Default {@link DataSourceRegistry} implementation. This implementation is used to register both local and XA data sources so that they can be resolved by extensions.
  */
 public class LocalDataSourceRegistry implements DataSourceRegistry {
-    private LocalTransactionContextManager manager;
```
Comment (jimmarino) [inline extensions/transaction/transaction-local/src/main/java/org/eclipse/dataspaceconnector/transaction/local/LocalDataSourceRegistry.java:28]: Hey @denisneuling can you explain why these changes were made to the LocalDataSourceRegistry?

**Review by jimmarino — CHANGES_REQUESTED** [review]

File: common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/ResultSetMapper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (denisneuling) [reply common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/ResultSetMapper.java]: @paullatzelsperger Alright, I'll give it a try, I'll introduce that `EdcPersistenceException` then. Where shall it come from? Shall I create a new class within `:extensions:sql:common` or directly in the `:spi` module?

File: extensions/transaction/transaction-local/src/main/java/org/eclipse/dataspaceconnector/transaction/local/LocalDataSourceRegistry.java (hunk 7 above)
Comment (jimmarino) [reply extensions/transaction/transaction-local/src/main/java/org/eclipse/dataspaceconnector/transaction/local/LocalDataSourceRegistry.java:28]: Sorry, I'm confused. Here is what I think should happen:

1. Each configuration of the ConnectionPool implementation registers exactly one DataSource that is backed by the pool during extension initialization.
2. This pooled DataSource implementation will be wrapped in a DataSourceResource and registered with the LocalTransactionContext, which acts as a transaction manager. 
3. DataSourceResource is threadsafe. When the getConnection() method is called, the DataSourceResource will delegate to the connection pool DataSource implementation if no connection is enlisted for the thread.  
4. The connection pool DataSource will lease and return a Connection from the pool. This connection instance will be pinned to the current thread by DataSourceResource.
5. Subsequent calls to return a connection within the same transaction context will return the same leased Connection.

Is there something I am missing that would cause this not to work?

Comment (denisneuling) [reply extensions/transaction/transaction-local/src/main/java/org/eclipse/dataspaceconnector/transaction/local/LocalDataSourceRegistry.java:28]: That's how it's working except that the `CommonsConnectionPoolServiceExtension` does rely on the `DataSourceRegistry` to provide it with already existing `DataSource`s to be wrapped.

1. The `CommonsConnectionPoolServiceExtension` [iterates through the config properties to find all configured `DataSource`s](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/499/files#diff-e879c6194f1780828137ba3c349f945e819d85cbb2bbaf488c5c3ac92102ce74R68-R80) 
2. Those `DataSource`s are [retrieved from the `DataSourceRegistry`, if found wrapped into a `ConnectionPoolDataSource` and again re-registered (overridden) at the `DataSourceRegistry`](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/499/files#diff-e879c6194f1780828137ba3c349f945e819d85cbb2bbaf488c5c3ac92102ce74R52-R56)
3. All `DataSource`s known to the `LocalDataSourceRegistry` become [registered at the `LocalTransactionContext` after `LocalTransactionExtension` intialization](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/499/files#diff-f37781325ab23fa4fa2ea8cf863e4c9a7c505cb54ff440a807b2f4aff8848e1bR51-R54)

If that's still unclear I'd be happy having a call to lead you through the code.

File: common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/ResultSetMapper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (paullatzelsperger) [reply common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/ResultSetMapper.java]: I think it can be in the `:spi` module, most likely `:spi:core-spi`. Ideally, all the other persistence extensions will use it as well then.

File: extensions/transaction/transaction-local/src/main/java/org/eclipse/dataspaceconnector/transaction/local/LocalDataSourceRegistry.java (hunk 7 above)
Comment (jimmarino) [reply extensions/transaction/transaction-local/src/main/java/org/eclipse/dataspaceconnector/transaction/local/LocalDataSourceRegistry.java:28]: I think I understand what the issue is. It looks as if you are trying to wrap *existing* connections in a pool that is then registered. That's not how it should work.

Instead, a pool implementation is responsible for creating a delegate DataSource from configuration, instantiating it, and giving it to the pool. The pool will then be registered in Step 1 that I outlined above.

Everything should then work fine and the changes made to the local transaction extension can be reverted.

File: common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/ResultSetMapper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (paullatzelsperger) [reply common/libraries/sql-lib/src/main/java/org/eclipse/dataspaceconnector/sql/ResultSetMapper.java]: If we want other persistence extensions (eg cosmos) to also throw it, spi is the common denominator, right?
If we don't, then it doesn't need to go into spi.

File: extensions/transaction/transaction-local/src/main/java/org/eclipse/dataspaceconnector/transaction/local/LocalDataSourceRegistry.java (hunk 7 above)
Comment (denisneuling) [reply extensions/transaction/transaction-local/src/main/java/org/eclipse/dataspaceconnector/transaction/local/LocalDataSourceRegistry.java:28]: Thought the only responsibility was be to provide pooling functionality without actually instantiating `Connection`s by calling the `DriverManager`. Seems I misunderstood the intention of that extension then...  Will then remove the `:extension:sql:postgresql` at the same adding the proposed functionality.

**Comment by denisneuling** [comment]
@jimmarino, changes to the registry and local tx manager have been reverted. 
The `CommonsConnectionPoolServiceExtension` utilizes the configuration to create and register pooled datasources to the datasource registry now. 
Also the `EdcPersistenceException` mentioned by Paul has been introduced. 
Is there anything else missing?

**Comment by jimmarino** [comment]
> @jimmarino, changes to the registry and local tx manager have been reverted. The `CommonsConnectionPoolServiceExtension` utilizes the configuration to create and register pooled datasources to the datasource registry now. Also the postgresql extension has been removed due to beeing redundant to what the commons-pool extension is capable of. Also the `EdcPersistenceException` mentioned by Paul has been introduced. Is there anything else missing?

Hi @denisneuling again sorry about the delay but I didn't see the notification. The only thing that needs to be done is to convert the checked exceptions to unchecked ones as mentioned by @paullatzelsperger.

**Comment by denisneuling** [comment]
@jimmarino Checked exceptions have already been substituted by `EdcPersistenceException`s everywhere useful / applicable with the last commit. If there aren't any other ToDo's let's get this merged please.

**Comment by paullatzelsperger** [comment]
> @jimmarino Checked exceptions have already been substituted by `EdcPersistenceException`s everywhere useful / applicable within my last commit. If there aren't any other ToDo's let's get this merged please.

The `ResultSetMapper` still declares an `Exception`

**Comment by paullatzelsperger** [comment]
> @jimmarino Checked exceptions have already been substituted by `EdcPersistenceException`s everywhere useful / applicable within my last commit. If there aren't any other ToDo's let's get this merged please.

The `ResultSetMapper` still declares an `Exception`

**Comment by denisneuling** [comment]
@paullatzelsperger The `ResultSetMapper` operates on `java.sql.ResultSet`s, which on each invocation could raise `java.sql.SQLException`s. Disallowing the `ResultSetMapper` to throw those exceptions would ruin it to be used as a lambda/function. Do we really want that? Also, if used in conjunction with the `SqlQueryExecutor`, thrown `java.sql.SQLException`s will be wrapped into `EdcPersistenceException`s - so no escalation of checked exception will occur.

**Comment by jimmarino** [comment]
> @paullatzelsperger The `ResultSetMapper` operates on `java.sql.ResultSet`s, which on each invocation could raise `java.sql.SQLException`s. Disallowing the `ResultSetMapper` to throw those exceptions would ruin it to be used as a lambda/function. Do we really want that? Also, if used in conjunction with the `SqlQueryExecutor`, thrown `java.lang.Exceptions`s will be wrapped into `EdcPersistenceException`s - so no escalation of checked exception will occur.

Having RowMapper raise SqlExceptions seems reasonable so that the shortened lambda syntax for single line expressions can be used. However, checked exceptions are raised in a number of other places.

Here's a diff that converts these to an unchecked exception:

https://gist.github.com/jimmarino/bdb8e7dd216cd786e14486bf522b5aa6

**Comment by paullatzelsperger** [comment]
@denisneuling once @jimmarino's changes are in, i'll approve.

**Comment by denisneuling** [comment]
@jimmarino Although I appreciate your effort, the `ArgumentHandler` is package private - so not visible to the outside - and it is just used by the `SqlQueryExecutor` which indeed wraps all `SQLException`s. 
Hiding `SQLException`s from the `ConnectionPool` and `ConnectionFactory` doesn't seem reasonable to me, but fine, the changes have been applied there to just get this merged.

**Review by paullatzelsperger — APPROVED** [review]
